### DISCUSSION TITLE
Get text from diagnostic for vimscript

### DISCUSSION THREAD

**Discussion by otavioschwanck** [neoclide/coc.nvim#3711] [Q&A] [answered]
Hi, 

at first, thanks for your attention.

I want to build a feature that get the text on the diagnostic, parse and insert at the end.  

Example:

![image](https://user-images.githubusercontent.com/9551316/158697381-234d40a7-0d64-407d-8b7d-ed1753484e66.png)

I want to pick the [Layout/IndentationConsistency] from the tooltip and insert at the end of the line with some text.  At final:

```ruby
before_action :populate_form, only: [:new_boleto, :create_boleto] # rubocop:disable Layout/IndentationConsistency
```

I have no clue of how to do this on vimscript.  Im a ex emacs user and have this on my configuration.

Thanks for your attetion!

**Comment by chemzqm** [comment]
`:h g:coc_last_float_win`, `:h winbufnr()`, `:h getbufline()`, `:h setline()` or `:h nvim_buf_set_text()` (neovim > =0.6.1)

**Reply by chemzqm** [reply]
But it's better to create a simple coc extension for that.

**Comment by chemzqm** [comment, marked as answer]
``` js
const {diagnosticManager, workspace, Position, commands, TextEdit} = require('coc.nvim')

exports.activate = async context => {
  let {nvim} = workspace
  context.subscriptions.push(commands.registerCommand('rubocop.comment', async () => {
    let [pos, line, bufnr] = await nvim.eval('[coc#util#cursor(),getline("."),bufnr("%")]')
    let diagnostics = await diagnosticManager.getCurrentDiagnostics()
    if (diagnostics.some(d => d.message.includes('Inconsistent indentation'))) {
      // add comment
      let doc = workspace.getDocument(bufnr)
      let edit = TextEdit.insert(Position.create(pos.line, line.length), '# rubocop:disable Layout/IndentationConsistency')
      await doc.applyEdits([edit])
    }
  }))
}
```
Save the javascript to a js file in your `~/.vim/coc-extensions` folder.

**Reply by otavioschwanck** [reply]
tried to put on ~/.config/coc/extensions, nothing happens =C

Where i put it?  How to execute it inside neovim after?

**Reply by chemzqm** [reply]
You need run command `:CocCommand rubocop.comment` or pick it from command list by `:CocCommand`

**Reply by chemzqm** [reply]
It would be better to use `languages.registerCodeActionProvider()` api, so it can be fixed by `<Plug>(coc-fix-current)`

**Reply by otavioschwanck** [reply]
> It would be better to use `languages.registerCodeActionProvider()` api, so it can be fixed by `<Plug>(coc-fix-current)`

i use neovim, where i put this JS file?

**Reply by otavioschwanck** [reply]
> It would be better to use `languages.registerCodeActionProvider()` api, so it can be fixed by `<Plug>(coc-fix-current)`

Just found already where to put, but the rubocop.comment has no action on the buffer.

My code:

```
const {diagnosticManager, workspace, Position, commands, TextEdit} = require('coc.nvim')

exports.activate = async context => {
  let {nvim} = workspace
  context.subscriptions.push(commands.registerCommand('rubocop.comment', async () => {
    let [pos, line, bufnr] = await nvim.eval('[coc#util#cursor(),getline("."),bufnr("%")]')
    let diagnostics = await diagnosticManager.getCurrentDiagnostics()
    if (diagnostics.some(d => d.message.includes('rubocop'))) {
      // add comment
      let doc = workspace.getDocument(bufnr)
      let edit = TextEdit.insert(Position.create(pos.line, line.length), diagnostics[0])
      await doc.applyEdits([edit])
    }
  }))
}
```

i think the textedit insert part is not working

**Reply by chemzqm** [reply]
Use `pos[0]` instead
```
      let edit = TextEdit.insert(Position.create(pos[0], line.length), '# rubocop:disable Layout/IndentationConsistency')
```

**Reply by otavioschwanck** [reply]
> .some(d => d.message.includes('rubocop'))

I ended doing that:

```
const {diagnosticManager, workspace, Position, commands, TextEdit} = require('coc.nvim')

exports.activate = async context => {
    let {nvim} = workspace
    context.subscriptions.push(commands.registerCommand('rubocop.insert', async () => {
        let [, line, ] = await nvim.eval('[coc#util#cursor(),getline("."),bufnr("%")]')

        let diagnostics = await diagnosticManager.getCurrentDiagnostics()

        if (diagnostics.length > 0) {
            let message = diagnostics[0].message.split(" ")[0]
            let parsedMessage = message.substr(1, message.length - 2)

            if (line.includes("# rubocop:disable")) {
                await nvim.call('setline', ['.', line + ", " + parsedMessage])
            } else {
                await nvim.call('setline', ['.', line + " # rubocop:disable " + parsedMessage])
            }
        }
    }))
}
``` 

Worked like a charm!

Thanks!
